amq.go

```diff
@@ -0,0 +1,120 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"github.com/Azure/go-amqp"
+	"github.com/prometheus/client_golang/prometheus"
+	"log"
+	"time"
+)
+
+var (
+	messagesSentGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "amq_messages_sent_success",
+	})
+	messagesReceivedGauge = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "amq_messages_received_success",
+	})
+)
+
+func init()  {
+	prometheus.MustRegister(messagesSentGauge)
+	prometheus.MustRegister(messagesReceivedGauge)
+}
+
+type AMQChecks struct {
+	address string
+	queueName string
+	sendTimeout time.Duration
+	interval time.Duration
+}
+
+func (a *AMQChecks) runForever() error {
+	// Create client
+	t := &tls.Config{InsecureSkipVerify: true}
+	opts := amqp.ConnTLSConfig(t)
+	client, err := amqp.Dial(a.address, opts)
+	if err != nil {
+		return fmt.Errorf("failed to connect to %s: %v", a.address, err)
+	}
+	defer client.Close()
+
+	// Create session
+	session, err := client.NewSession()
+	if err != nil {
+		return fmt.Errorf("failed to create new session: %v", err)
+	}
+
+	// Create a sender
+	sender, err := session.NewSender(
+		amqp.LinkTargetAddress(a.queueName),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create a new sender: %v", err)
+	}
+
+	// Create a receiver
+	receiver, err := session.NewReceiver(
+		amqp.LinkSourceAddress(a.queueName),
+		amqp.LinkCredit(10),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create a new receiver: %v", err)
+	}
+
+	// Start receiving messages
+	go func(receiver *amqp.Receiver) {
+		a.receiveMessages(receiver)
+	}(receiver)
+
+	//start send messages
+	a.sendMessages(sender)
```

Comment:
What if the message is sent successfully but is not received, shouldn't we report an error if it happened? Or we are leaving this problem to who visualize the data to check if the number of sent equals the number of received messages?

Reply:
Both `sendMessage` and `receiveMessage` will only set the gauge metric values to be 1 when they finished successfully, all the other scenarios they will be set to 0. This should allow us to easily spot any downtime for send and receive functions.

In the case you described, it will be one gauge value is 1 and the other is 0, and you should be able to see it from the graph.